### PR TITLE
[1131] Alternative implementation of site status validation

### PR DESCRIPTION
**Pull request by benilovj**
### Context
An alternative approach to #196.

### Changes proposed in this pull request
Add validation to site statuses and fix the specs/factories that break as a result.

### CODE REVIEW

**Comment by benilovj** [comment]
I've still got to:

1. add specs for the validation implementation itself
2. fix the failures in `spec/models/course_spec.rb` where site statuses are being created before the course.

File: app/models/site_status.rb [hunk 1]
```diff
@@ -43,4 +49,20 @@ class SiteStatus < ApplicationRecord
   }
   scope :with_vacancies, -> { where.not(vac_status: :no_vacancies) }
   scope :open_for_applications, -> { findable.applications_being_accepted_now.with_vacancies }
+
+private
+
+  def vac_status_consistent_with_course_study_mode?
+    if no_vacancies?
```
Comment (dankmitchell) [inline app/models/site_status.rb]: A case statement might be easier to read...

Comment (benilovj) [reply app/models/site_status.rb]: The linter objects to case statements without a subject, unfortunately.

In any case, we think we're gonna back out the actual validation from this PR and just take the spec changes forward, so #196 will have the actual code.

File: app/models/site_status.rb [hunk 2]
```diff
@@ -14,6 +14,12 @@
 class SiteStatus < ApplicationRecord
   self.table_name = "course_site"
 
+  validate if: Proc.new { |s| s.course&.study_mode.present? } do
+    unless vac_status_consistent_with_course_study_mode?
+      errors.add(:vac_status, "(#{vac_status}) must be consistent with course study mode #{course.study_mode}")
+    end
+  end
```
Comment (misaka) [inline app/models/site_status.rb]: I was trying to say earlier before my daughter a bit rudely interrupted 🤢 ... it'd be nice to move the body of this to a method. I know it already uses `vac_status_consistent_with_course_study_mode?`, which is good, but I think it be cleaner to keep this part of the model free of code (esp multi-line block) and just have declarations here.

Worth considering if you're going to move this to the other PR.

File: spec/factories/site_statuses.rb [hunk 3]
```diff
@@ -13,11 +13,15 @@
 
 FactoryBot.define do
   factory :site_status do
-    association(:course)
+    association :course, study_mode: :full_time_or_part_time
     association(:site)
     publish { 'N' }
     vac_status { :full_time_vacancies }
```
Comment (misaka) [inline spec/factories/site_statuses.rb]: This status isn't always going to be valid, why not do the `with_any_vacancy` logic by default?

File: app/models/site_status.rb (hunk 1 above)
Comment (dankmitchell) [reply app/models/site_status.rb]: @benilovj could you just use `case vac_status: no_vacancies` instead?

Comment (misaka) [reply app/models/site_status.rb]: Dunno if it's clear, but I think @dankmitchell means:

```
case vac_status
when 'no_vacancies'
  ...
```

Comment (dankmitchell) [reply app/models/site_status.rb]: Thanks @misaka 💯

Comment (benilovj) [reply app/models/site_status.rb]: Fixed.

File: app/models/site_status.rb (hunk 2 above)
Comment (benilovj) [reply app/models/site_status.rb]: Done.

File: spec/factories/site_statuses.rb (hunk 3 above)
Comment (benilovj) [reply spec/factories/site_statuses.rb]: I think a static default value is simpler as it bypasses the `after(:build)` hook and is therefore less magic. Plus a bunch of stuff breaks if I make `with_any_vacancy` the default, so it didn't feel like the right time to switch this.

**Comment by benilovj** [comment]
I can leave the validation implementation in or take it out – happy to be steered by the reviewers.

File: app/models/site_status.rb [hunk 4]
```diff
@@ -43,4 +46,27 @@ class SiteStatus < ApplicationRecord
...
+
+private
+
+  def vac_status_must_be_consistent_with_course_study_mode
+    unless vac_status_consistent_with_course_study_mode?
+      errors.add(:vac_status, "(#{vac_status}) must be consistent with course study mode #{course.study_mode}")
+    end
+  end
+
+  def vac_status_consistent_with_course_study_mode?
```
Comment (dankmitchell) [inline app/models/site_status.rb]: Nice! We’ve got this same method on my frontend PR so good to have here instead 👌

File: app/models/site_status.rb [hunk 5]
```diff
@@ -43,4 +46,27 @@ class SiteStatus < ApplicationRecord
   }
   scope :with_vacancies, -> { where.not(vac_status: :no_vacancies) }
   scope :open_for_applications, -> { findable.applications_being_accepted_now.with_vacancies }
+
+private
+
+  def vac_status_must_be_consistent_with_course_study_mode
+    unless vac_status_consistent_with_course_study_mode?
+      errors.add(:vac_status, "(#{vac_status}) must be consistent with course study mode #{course.study_mode}")
```
Comment (defong) [inline app/models/site_status.rb]: Think it's needs message

errors.add(:vac_status,
                 :invalid,
                 message:

File: app/models/site_status.rb (hunk 4 above)
Comment (heidar) [reply app/models/site_status.rb]: Let's keep this for consistency then

File: app/models/site_status.rb (hunk 5 above)
Comment (heidar) [reply app/models/site_status.rb]: As it is works fine, but you highlighted that the tests don't check for the validation error, which is really important!

**Comment by defong** [comment]
[Site status validation](https://trello.com/c/fKdscaJb/1187-site-status-validation)

**Pull request merged**